The master proxy was too slow to erase a GRV budget deficit if no GRV requests were coming in.

ajbeamon opened the pull request.

ajbeamon commented:
I'm adding some additional logic to clamp the budget deficit to this PR, that should be coming shortly.

ajbeamon commented:
I've added the new logic. The knob values were chosen somewhat arbitrarily, so let me know if you think they should change.

dongxinEric commented on fdbserver/MasterProxyServer.actor.cpp:
```diff
@@ -1108,15 +1105,25 @@ struct TransactionRateInfo {
 	void reset(double elapsed) {
-		limit = std::min(0.0,limit) + std::min(rate * elapsed, SERVER_KNOBS->START_TRANSACTION_MAX_TRANSACTIONS_TO_START);
+		limit = std::min(0.0, limit) + rate * elapsed; // Adjust the limit based on the full elapsed interval in order to properly erase a deficit
+		limit = std::min(limit, rate * SERVER_KNOBS->START_TRANSACTION_BATCH_INTERVAL_MAX); // Don't allow the rate to exceed what would be allowed in the maximum batch interval
```
Is this necessary? It seems like `elapsed` will not be larger than `SERVER_KNOBS->START_TRANSACTION_BATCH_INTERVAL_MAX` and thus `std::min(0.0, limit) + rate * elapsed` will always be a value smaller or equal to `rate * SERVER_KNOBS->START_TRANSACTION_BATCH_INTERVAL_MAX`.

dongxinEric replied:
Oh never mind, I didn't see the change below.

ajbeamon commented:
@fdb-build test CTest please

ajbeamon commented:
I've reverted the code that clamps the deficit for now because I think there are some flaws both with the current approach and the new one I had implemented. My intent is to try to address those in a future PR.
